rust/processor/src/worker.rs

```diff
@@ -205,6 +209,74 @@ impl Worker {
+        tokio::spawn(async move {
+            info!(
+                processor_name = processor_name,
+                service_type = PROCESSOR_SERVICE_TYPE,
+                "[Parser] Starting gap detector task",
+            );
+
+            let mut gap_detector = GapDetector::new(starting_version);
+            let mut num_batches_processed_without_gap = 0;
+
+            loop {
+                let result = match gap_detector_receiver.recv().await {
+                    Some(result) => result,
+                    None => {
+                        info!(
+                            processor_name,
+                            service_type = PROCESSOR_SERVICE_TYPE,
+                            "[Parser] Gap detector channel has been closed",
+                        );
+                        return;
+                    },
+                };
+
+                match gap_detector.process_versions(result) {
+                    Ok(GapDetectorResult::GapDetected { gap_start_version }) => {
+                        error!(
+                            processor_name,
+                            gap_start_version,
+                            "[Parser] Processed {GAP_DETECTION_BATCH_COUNT} batches with a gap. Panicking."
+                        );
+                        panic!("[Parser] Processed {GAP_DETECTION_BATCH_COUNT} batches with a gap. Panicking.");
+                    },
+                    Ok(GapDetectorResult::NoGapDetected {
+                        last_success_batch,
+                        num_batches_processed_without_gap: num_batches,
+                    }) => {
+                        num_batches_processed_without_gap += num_batches;
+                        if num_batches_processed_without_gap >= PROCESSOR_STATUS_UPDATE_BATCH_COUNT
+                        {
+                            // gap_detector.update_last_processed_version(last_success_batch, None).await.unwrap();
+                            let last_success_batch = last_success_batch.unwrap();
+                            processor_clone
+                                .update_last_processed_version(
+                                    last_success_batch.end_version,
+                                    last_success_batch.last_transaction_timstamp.clone(),
+                                )
+                                .await
+                                .unwrap();
+                            num_batches_processed_without_gap = 0;
+                        }
+                    },
+                    Err(e) => {
+                        error!(
+                            processor_name,
+                            service_type = PROCESSOR_SERVICE_TYPE,
+                            error = ?e,
+                            "[Parser] Gap detector task has panicked"
+                        );
+                        panic!();
+                    },
+                }
+            }
+        });
```

Comment:
can you move this to it's own function? will make this section much easier to reason through